streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java

```diff
@@ -361,6 +363,8 @@ void updateRestored(final Collection<TopicPartition> restored) {
         }
         if (allTasksRunning()) {
             restoredPartitions.clear();
+            restoringByPartition.clear();
+            restoring.clear();
```

Comment:
Why is this necessary? In the `for` loop above, each partition that finished restoring should be removed vai `it.remove();` -- hence, if `allTasksRunning()` return `true` `restoring` should be empty ?

Similar for `restoringByPartition` given the two new lines this PR adds.

Reply:
Btw: we maintain state quite redundantly (what is a general issue). Would it make sense to extract a helper class that wraps
```
private final Map<TaskId, StreamTask> restoring = new HashMap<>();
private final Set<TopicPartition> restoredPartitions = new HashSet<>();
private final Map<TopicPartition, StreamTask> restoringByPartition = new HashMap<>();
```

and we only modify the state via the helper? The helper class can insure that all three data structures don't diverge. (Also ok to do in a follow up cleanup PR)

Reply:
Ack to the first point, we shouldn't need to clear these here. I also completely agree with the second comment, but we want to limit these PRs to just the necessary fixes for now -- we've been discussing some other large-scale refactoring and cleanup of this and related classes on other PRs and tried to drop most of the thoughts on this ticket, if you have anything else to add: https://issues.apache.org/jira/browse/KAFKA-9113